### PR TITLE
Add canonicalization of a GEQ -> tosa.select to tosa.clamp

### PR DESCRIPTION
**Pull request by jorickert**
This appears in the lowering of onnx to tosa

### CODE REVIEW

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 1]
```diff
@@ -114,15 +145,63 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+    return success();
+  }
+};
+
+// This pattern canonicalizaties the following pattern:
+// %0 = tosa.greater_equal(input, 0)
+// %1 = tosa.select(%0, input, 0)
+// to tosa.clamp{min = 0, max = max}(input)
+// This pattern occurs in decompositions of LeakyReLU/PReLU with an alpha
+// of zero
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:168]: Can we make this canonicalization for any splat value? I believe it doesn't need to be specifically for 0.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 2]
```diff
@@ -114,15 +145,63 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+struct SelectToClampOptimization : public OpRewritePattern<tosa::SelectOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(tosa::SelectOp op,
+                                PatternRewriter &rewriter) const override {
+    auto onFalse = op.getOnFalse();
+    DenseElementsAttr onFalseAttr;
+    if (!matchPattern(onFalse, m_Constant(&onFalseAttr)) ||
+        !isSplatZero(onFalse.getType().getElementType(), onFalseAttr)) {
+      return failure();
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:177]: can we change these to return rewrite.notifymatchfailure with a message.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 3]
```diff
@@ -114,15 +145,63 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+    }
+    auto input = op.getOnTrue();
+    if (input.getDefiningOp() != geq.getInput1().getDefiningOp()) {
+      return failure();
+    }
+    rewriter.replaceOpWithNewOp<tosa::ClampOp>(
+        op, op.getType(), input, rewriter.getI64IntegerAttr(0),
+        rewriter.getI64IntegerAttr(std::numeric_limits<int64_t>::max()),
+        rewriter.getF32FloatAttr(0.0f),
+        rewriter.getF32FloatAttr(std::numeric_limits<float>::max()));
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:196]: In the current implementation, you have checked this select (x>=0, x , 0) and you are setting the min attr for a clip. You can also check this select. select(x>=0, 0, x) [0 is the on true] and you can set the max.

I believe you can generalize this canonicalization to (x>=y, y, x) where y is a splat const and you have to track if y is on true or false condition and set max or min correspondingly.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp (hunk 1 above)
Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp]: Done

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp (hunk 2 above)
Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp]: Done

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp (hunk 3 above)
Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp]: Done

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 4]
```diff
@@ -114,15 +114,159 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+        !geqIn2Attr.isSplat()) {
+      return rewriter.notifyMatchFailure(
+          op, "RHS of predicate GreaterEqualOp is not a SplatValue");
+    }
+    auto inputElementType = geqIn2Attr.getElementType();
+
+    auto onFalse = op.getOnFalse();
+    auto onTrue = op.getOnTrue();
+
+    auto replaceSelectWithClamp = [&](int64_t intMin, int64_t intMax,
```
Comment (ehsan-toosi) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:164]: You can make this template that template parameter is either IntegerType or FloatType and it will set automatically to the attributes you need to set. You can also make the argument optional so that if it's not provided, you automatically set them in this function.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 5]
```diff
@@ -114,15 +114,159 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+    if (matchPattern(onFalse, m_Constant(&onFalseAttr)) &&
+        isCompatibleSplat(onFalseAttr, geqIn2Attr) &&
+        onTrue.getDefiningOp() == geq.getInput1().getDefiningOp()) {
+      if (isa<IntegerType>(inputElementType)) {
+        auto splatValue = geqIn2Attr.getSplatValue<APInt>().tryZExtValue();
+        if (!splatValue) {
+          return rewriter.notifyMatchFailure(
+              op, "Can not rewrite as values do not fit into 64 bits");
+        }
+        replaceSelectWithClamp(*splatValue, std::numeric_limits<int64_t>::max(),
```
[the remark this answers is not in the document]
Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:211]: This is not needed, as the int type is always int64_t for the clamp. For the float the type can vary.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 6]
```diff
@@ -114,15 +114,159 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+        }
+        replaceSelectWithClamp(std::numeric_limits<int64_t>::min(), *splatValue,
+                               rewriter.getF32FloatAttr(0.0f /*Unused*/),
+                               rewriter.getF32FloatAttr(0.0f /*Unused*/));
+        return success();
+      }
+      if (isa<FloatType>(inputElementType)) {
+        auto splatValue = geqIn2Attr.getSplatValue<APFloat>();
+        replaceSelectWithClamp(
+            0 /*Unused*/, 0 /*Unused*/,
```
[the remark this answers is not in the document]
Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:252]: There is no value specified. Also the tosa spec and tosa-mlir implementation are quite different.
Personally I think the 0, 0 as default makes the mlir more readable

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 7]
```diff
@@ -114,15 +114,148 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+      return failure();
+    rewriter.modifyOpInPlace(op, [&]() {
+      op.getOperation()->setOperands(
+          {notOp.getInput1(), op.getOnFalse(), op.getOnTrue()});
+    });
+    return success();
+  }
+};
+
+// This canonicalizaties the following patterns:
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:132]:
```suggestion
// This canonicalizes the following patterns:
```

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 8]
```diff
@@ -114,15 +114,148 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+        !geqIn2Attr.isSplat()) {
+      return rewriter.notifyMatchFailure(
+          op, "RHS of predicate GreaterEqualOp is not a SplatValue");
+    }
+    auto isCompatibleSplat = [](DenseElementsAttr a,
+                                DenseElementsAttr b) -> bool {
+      if (!a.isSplat() || !b.isSplat()) {
+        return false;
+      }
+      if (a.getElementType() != b.getElementType()) {
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:164]: is this really necessary? I think tosa expects that the both types are always the same, and this should only be triggered in invalid IR. Maybe it can be removed.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 9]
```diff
@@ -114,15 +114,148 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+        return false;
+      }
+      auto aType = llvm::dyn_cast<RankedTensorType>(a.getType());
+      auto bType = llvm::dyn_cast<RankedTensorType>(a.getType());
+      if (!(aType == bType ||
+            (aType.getRank() == 1 && aType.getDimSize(0) == 1) ||
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:170]: I think the spec also expects that both ranks are the same, i.e., if one is 4 the other is also 4.
References here: https://www.mlplatform.org/tosa/tosa_spec.html#_select and https://www.mlplatform.org/tosa/tosa_spec.html#_broadcast_helpers
I don't think we need this. WDYT?

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 10]
```diff
@@ -114,15 +114,148 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+      int64_t clampIntMax = std::numeric_limits<int64_t>::max();
+      FloatAttr clampFloatMin;
+      FloatAttr clampFloatMax;
+      if (isa<IntegerType>(inputElementType)) {
+        auto splatValue = geqIn2Attr.getSplatValue<APInt>().trySExtValue();
+        if (!splatValue) {
+          return rewriter.notifyMatchFailure(
+              op, "Can not rewrite as values do not fit into 64 bits");
+        }
+        clampFloatMin =
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:220]: haven't we checked that already?

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 11]
```diff
@@ -114,15 +114,148 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+    // %0 = tosa.greater_equal(input, cmp)
+    // %1 = tosa.select(%0, cmp, input)
+    // to tosa.clamp{min = input, max = cmp}(input)
+    // Predicate: geq.input2 == select.onTrue AND geq.input1 == select.onFalse
+    const bool isCaseTwo =
+        !isCaseOne && matchPattern(onTrue, m_Constant(&onTrueAttr)) &&
+        isCompatibleSplat(onTrueAttr, geqIn2Attr) &&
+        onFalse.getDefiningOp() == geq.getInput1().getDefiningOp();
+
+    if (isCaseOne || isCaseTwo) {
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:208]: maybe invert the logic here to avoid a big if statement.
```suggestion
    if (!isCaseOne && !isCaseTwo) {
    return rewriter.notifyMatchFailure();
    }
```

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 12]
```diff
@@ -114,15 +114,148 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
-LogicalResult SelectOp::canonicalize(SelectOp op, PatternRewriter &rewriter) {
-  auto notOp = op.getPred().getDefiningOp<tosa::LogicalNotOp>();
-  if (!notOp)
-    return failure();
-  rewriter.modifyOpInPlace(op, [&]() {
-    op.getOperation()->setOperands(
-        {notOp.getInput1(), op.getOnFalse(), op.getOnTrue()});
-  });
-  return success();
+struct SelectLogicalNotOptimization : public OpRewritePattern<tosa::SelectOp> {
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:117]: do we have a test for this?

Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:117]: This is an existing pattern that I needed to move into its own pattern. The existing test for it is: `select_not_pred``

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 13]
```diff
@@ -114,15 +114,159 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
...
+                               rewriter.getF32FloatAttr(0.0f /*Unused*/));
+        return success();
+      }
+      if (isa<FloatType>(inputElementType)) {
+        auto splatValue = geqIn2Attr.getSplatValue<APFloat>();
+        replaceSelectWithClamp(
+            0 /*Unused*/, 0 /*Unused*/,
+            rewriter.getFloatAttr(
+                inputElementType,
+                APFloat::getSmallest(splatValue.getSemantics())),
```
[the remark this answers is not in the document]
Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp]: @ttjost I also think using inf is better, as llvm can in some cases optimize a clamp with  max/min = inf to a no-op, but max/min = largest prevents this optimization

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp (hunk 10 above)
Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:209]: No. We only check that it is an int or float. Clamp does only support int64, (but arbitrary float types), so we need to check if we can represent the clamp value as int

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 14]
```diff
@@ -73,56 +73,187 @@ struct SqrtReciprocalOptimization : public OpRewritePattern<tosa::PowOp> {
...
+          op, "select does not match GEQ + select -> clamp pattern");
+    }
+
+    const auto inputElementType = geqIn2Attr.getElementType();
+    int64_t clampIntMin = std::numeric_limits<int64_t>::min();
+    int64_t clampIntMax = std::numeric_limits<int64_t>::max();
+    FloatAttr clampFloatMin;
+    FloatAttr clampFloatMax;
+    if (isa<IntegerType>(inputElementType)) {
+      auto splatValue = geqIn2Attr.getSplatValue<APInt>().trySExtValue();
```
Comment (cmcgirr-amd) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:218]: Our fork of LLVM allows for unsigned types (Tosa_Tensor def is not strictly following the spec) here as well could we guard against it? Or add the handling for unsigned. Either one.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 15]
```diff
@@ -73,56 +73,187 @@ struct SqrtReciprocalOptimization : public OpRewritePattern<tosa::PowOp> {
                                 PatternRewriter &rewriter) const override {
     // Check that the PowOp has a single user
     if (!op->hasOneUse())
-      return rewriter.notifyMatchFailure(op, "pow operator has more than one user");
+      return rewriter.notifyMatchFailure(op,
```
Comment (cmcgirr-amd) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:76]: nit: I see a lot of formatting issues that are unrelated to this change. I like to use `git add -p` to add chunks of my diff at a time selecting which ones I want to be included in the commits. Just to make sure I have the changes I want.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 16]
```diff
@@ -73,56 +73,187 @@ struct SqrtReciprocalOptimization : public OpRewritePattern<tosa::PowOp> {
...
+      if (!a.isSplat() || !b.isSplat()) {
+        return false;
+      }
+      if (llvm::isa<IntegerType>(a.getElementType())) {
+        return a.getSplatValue<APInt>() == b.getSplatValue<APInt>();
+      }
+      if (llvm::isa<FloatType>(a.getElementType())) {
+        return a.getSplatValue<APFloat>() == b.getSplatValue<APFloat>();
+      }
+      return false; // Only int and float types are supported
```
Comment (cmcgirr-amd) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:179]: Can you add a test cases that triggers this kind of scenario where we do not have splats

**Review by cmcgirr-amd — APPROVED** [review]

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp (hunk 10 above)
Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp]: Added check for unsiged ints

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp (hunk 14 above)
Comment (jorickert) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp]: Added handling for unsigned with less than 63 bits
